source/extensions/filters/http/grpc_field_extraction/filter.cc

```diff
@@ -0,0 +1,229 @@
+                   grpc_status, error_msg);
+  decoder_callbacks_->sendLocalReply(
+      static_cast<Envoy::Http::Code>(Utility::grpcToHttpStatus(grpc_status)), error_msg, nullptr,
+      grpc_status, rc_detail);
+}
+
+Envoy::Http::FilterHeadersStatus Filter::decodeHeaders(Envoy::Http::RequestHeaderMap& headers,
+                                                       bool) {
+  if (!Grpc::Common::isGrpcRequestHeaders(headers)) {
+    ENVOY_STREAM_LOG(debug,
+                     "Request isn't gRPC as its headers don't have application/grpc content-type. "
+                     "Passed through the request "
+                     "without extraction.",
+                     *decoder_callbacks_);
+    return Http::FilterHeadersStatus::Continue;
+  }
+
+  // Grpc::Common::isGrpcRequestHeaders above already ensures the existence of ":path" header.
+  auto proto_path = GrpcPathToProtoPath(headers.Path()->value().getStringView());
+  if (!proto_path.ok()) {
+    ENVOY_STREAM_LOG(info, "failed to convert gRPC path to protobuf path: {}", *decoder_callbacks_,
+                     proto_path.status().ToString());
+    auto& status = proto_path.status();
+    rejectRequest(status.raw_code(), status.message(),
+                  generateRcDetails(kRcDetailFilterGrpcFieldExtraction,
+                                    absl::StatusCodeToString(status.code()),
+                                    kRcDetailErrorTypeBadRequest));
+    return Http::FilterHeadersStatus::StopIteration;
+  }
+
+  auto* extractor = filter_config_.FindExtractor(*proto_path);
+  if (!extractor) {
+    ENVOY_STREAM_LOG(debug, "gRPC method `{}` isn't configured for field extraction",
+                     *decoder_callbacks_, *proto_path);
+    return Http::FilterHeadersStatus::Continue;
+  }
+
+  extractor_ = extractor;
+  request_msg_converter_ = std::make_unique<MessageConverter>(
+      std::make_unique<std::function<std::unique_ptr<Protobuf::field_extraction::MessageData>()>>(
+          []() { return std::make_unique<Protobuf::field_extraction::CordMessageData>(); }),
+      decoder_callbacks_->decoderBufferLimit());
+
+  return Envoy::Http::FilterHeadersStatus::StopIteration;
+}
+
+Envoy::Http::FilterDataStatus Filter::decodeData(Envoy::Buffer::Instance& data, bool end_stream) {
+  ENVOY_STREAM_LOG(debug, "decodeData: data size={} end_stream={}", *decoder_callbacks_,
+                   data.length(), end_stream);
+  if (!extractor_ || extraction_done_) {
+    return Envoy::Http::FilterDataStatus::Continue;
+  }
+
+  if (auto status = handleDecodeData(data, end_stream); !status.got_messages) {
+    return status.filter_status;
+  }
+
+  return Envoy::Http::FilterDataStatus::Continue;
+}
+
+Filter::HandleDecodeDataStatus Filter::handleDecodeData(Envoy::Buffer::Instance& data,
+                                                        bool end_stream) {
+  ABSL_DCHECK(extractor_ && request_msg_converter_);
```

Comment:
Envoy has its own macros for assertions and debug checks. Try `ENVOY_BUG` or `RELEASE_ASSERT` maybe

Reply:
DONE.